### ISSUE TITLE
No annex uuid after "git annex initremote"

### ISSUE THREAD

**Issue by anikfal** [datalad/datalad#7437]
**In a HPC server:**
`git annex initremote myserver type=git autoenable=true location=ssh://nikfal1@jurca.fz-juelich.de:/p/largedata2/detectdata/newdatalad`

**Error:**
```
initremote myserver 
  Unable to parse git config from ssh://nikfal1@jurca.fz-juelich.de:/p/largedata2/detectdata/newdatalad

git-annex: git repository does not have an annex uuid
failed
initremote: 1 failed
```
I need this command, so I can do `datalad get` without applying `git remote add repo_name` after cloning from GitLab.
I had applied this command on another datalad dataset on the server and it was successful.

**git config:**
```
[core]
        repositoryformatversion = 0
        filemode = true
        bare = false
        logallrefupdates = true
[annex]
        uuid = 7f07d38e-09f0-4896-abb2-be0d32bc2e19
        version = 8
[filter "annex"]
        smudge = git-annex smudge -- %f
        clean = git-annex smudge --clean -- %f
[submodule "inputs"]
        active = true
        url = ./inputs
[remote "gitlabbb"]
        url = https://gitlab.jsc.fz-juelich.de/nikfal1/temp_datalad.git
        fetch = +refs/heads/*:refs/remotes/gitlabbb/*
        annex-ignore = true
```

**Comment by mslw** [comment]
We had a chat with @anikfal about this issue during an office hour, which allowed me to take another look afterwards. Here's what we learned.

### Initial problem

1. `git annex initremote myremote type=git location=ssh://... autoenable=true` attempts to connect to the given location to check the repository located there. What it looks for, specifically, is the annex uuid.
2. The URL in the original issue has a typo, but we confirmed the same behaviour without the typo.
3. In this case, we were running `git annex initremote` from the machine pointed at by the ssh url. AFAIK it is generally possible for a machine to connect "onto itself" via SSH, but in this case we were refused (silently, but indications in debug mode) the connection due to missing ssh key (the machine apparently only allows login via ssh keys, and the key wasn't present on the machine we were already logged into).
4. We were able to run the same `initremote` command from a local machine (which had the right ssh key).

Note _ad_ 3: with SSH agent forwarding it should be possible to log into the ssh machine, and have it do the "onto itself" thing.


### More general problem

I think this:

> I need this command, so I can do datalad get without applying git remote add repo_name after cloning from GitLab.
I had applied this command on another datalad dataset on the server and it was successful.

can be rephrased as:

I have an dataset on a machine that I access via SSH. I added a GitLab (GitHub-like) sibling to that dataset, and pushed. I want clones made from GitLab to other machines to automatically enable that dataset location for getting via SSH.

As you suggested, the type=git special remote *is* the right tool for the job (sorry for suggesting otherwise in our chat). Although (see above) it should be possible to `initremote` on that SSH machine (assuming login abilities, i.e. ssh agent forwarding or typed password), let's go through GitLab (partially for illustration, partially because I only remembered agent forwarding after typing in my response).

tl;dr: within `initremote`, set `autoenabled=true` if you want the remote to be enabled automatically

Example below.

We log in via SSH to our "remote" machine, and create a dataset with an annexed (by default) file:
```
datalad create foo
echo "1 2 3" > file.dat
datalad save
```
Then we add a GitLab sibling (could use `create-sibling-gitlab` - but I didn't have a token at hand):
```
datalad siblings add -s gitlab git@jugit.fz-juelich.de:m.szczepanik/convention-plane.git
datalad push --to gitlab
```

then we `logout` and back in our "local" machine we clone from GitLab:

```
datalad clone git@jugit.fz-juelich.de:m.szczepanik/convention-plane.git
cd convention-plane
```

Now a `get` operation would error, but our local clone kind of knows about the remote location - except the URL recorded by git-annex isn't very actionable for our local clone:

``` console
$ git cat-file -p git-annex:uuid.log
3625e93b-49d4-42a8-a287-76d4aa9a7857 mszczepanik@juseless:~/officehour/foo timestamp=1687878569.025021627s
f79c457a-af58-45e7-b00f-22124a87b803 mszczepanik@bnbnb64:/tmp/convention-plane timestamp=1687878934.862553479s
```

We can create an *autoenabled* type=git special remote (note: autoenabled is optional, but very important for what happens next):

``` console
$ git annex initremote sshserver type=git autoenable=true location=ssh://juseless.inm7.de/home/mszczepanik/officehour/foo
initremote sshserver ok
(recording state in git...)
```

Because git-annex checked the repository at the given location, it knows it has the annex UUID that we already knew (I must admit that surprised me initially, I thought we would need to specify `--sameas`, but apparently not):

``` console
$ git cat-file -p git-annex:uuid.log
3625e93b-49d4-42a8-a287-76d4aa9a7857 sshserver timestamp=1687879131.194128108s
f79c457a-af58-45e7-b00f-22124a87b803 mszczepanik@bnbnb64:/tmp/convention-plane timestamp=1687878934.862553479s

$ git cat-file -p git-annex:remote.log
3625e93b-49d4-42a8-a287-76d4aa9a7857 autoenable=true location=ssh://juseless.inm7.de/home/mszczepanik/officehour/foo name=sshserver type=git timestamp=1687879131.195376448s
```

Let's push to GitLab and make another clone (optionally: first declare the current clone dead if intending to remove it) - this should send the updated remote configuration to GitLab:

``` console
$ git annex dead here
$ datalad push --to origin
```

To test whether everything worked, we can remove this clone, and then clone from GitLab into a new location and get the file (skipping all outputs but the last):

``` console
$ cd ..
$ datalad drop --what all -d convention-plane
$ datalad clone git@jugit.fz-juelich.de:m.szczepanik/convention-plane.git new-convention-plane
$ cd new-convention-plane
$ datalad get file.dat
get(ok): file.dat (file) [from sshserver...]
```

Note: had we not configured our special remote to be autoenabled, after cloning we should see an INFO message "access to 1 dataset sibling sshserver not auto-enabled, enable with (...)", and we could indeed enable the remote in the dataset (`git annex initremote sshserver`) - this may be the preferred way of action if you don't expect having to enable the remote for all clones.
